src/ert/ensemble_evaluator/evaluator_async.py

```diff
@@ -0,0 +1,471 @@
+            set_handler(e_type, f)
+
+        while True:
+            batch: List[
+                Tuple[Callable[[List[CloudEvent]], Awaitable[None]], CloudEvent]
+            ] = []
+            start_time = asyncio.get_event_loop().time()
+            while len(batch) < 500 and asyncio.get_event_loop().time() - start_time < 2:
+                try:
+                    event = await asyncio.wait_for(self._events.get(), timeout=0.1)
+                    function = event_handler[event["type"]]
+                    batch.append((function, event))
+                    self._events.task_done()
+                except asyncio.TimeoutError:
+                    continue
+            await self._processing_queue.put(batch)
+
+    async def _fm_handler(self, events: List[CloudEvent]) -> None:
+        await self._append_message(self.ensemble.update_snapshot(events))
+
+    async def _started_handler(self, events: List[CloudEvent]) -> None:
+        if self.ensemble.status != ENSEMBLE_STATE_FAILED:
+            await self._append_message(self.ensemble.update_snapshot(events))
+
+    async def _stopped_handler(self, events: List[CloudEvent]) -> None:
+        if self.ensemble.status != ENSEMBLE_STATE_FAILED:
+            self._result = events[0].data  # normal termination
+            max_memory_usage = -1
+            for job in self.ensemble.snapshot.get_all_forward_models().values():
+                memory_usage = job.max_memory_usage or "-1"
+                if int(memory_usage) > max_memory_usage:
+                    max_memory_usage = int(memory_usage)
+            logger.info(
+                f"Ensemble ran with maximum memory usage for a single realization job: {max_memory_usage}"
+            )
+            await self._append_message(self.ensemble.update_snapshot(events))
+
+    async def _cancelled_handler(self, events: List[CloudEvent]) -> None:
+        if self.ensemble.status != ENSEMBLE_STATE_FAILED:
+            await self._append_message(self.ensemble.update_snapshot(events))
+            await self._stop()
+
+    async def _failed_handler(self, events: List[CloudEvent]) -> None:
+        if self.ensemble.status not in (
+            ENSEMBLE_STATE_STOPPED,
+            ENSEMBLE_STATE_CANCELLED,
+        ):
+            # if list is empty this call is not triggered by an
+            # event, but as a consequence of some bad state
+            # create a fake event because that's currently the only
+            # api for setting state in the ensemble
+            if len(events) == 0:
+                events = [await self._create_cloud_event(EVTYPE_ENSEMBLE_FAILED)]
+            await self._append_message(self.ensemble.update_snapshot(events))
+            await self._signal_cancel()  # let ensemble know it should stop
+
+    @property
+    def config(self) -> EvaluatorServerConfig:
+        return self._config
+
+    @property
+    def ensemble(self) -> Ensemble:
+        return self._ensemble
+
+    async def _create_cloud_event(
```

Comment:
What in this method is `async`?

Reply:
you are right, this can go without async :+1: